man/openipmigui.1
.TH openipmigui 1 05/13/03 OpenIPMI "GUI interface to an IPMI system"

.SH NAME
openipmigui \- GUI interface to an IPMI system

.SH SYNOPSIS
.B openipmigui
.RB [ \-\-dmsg ]
.RB [ \-\-drawmsg ]
.RB [ \-\-glib12 ]
.RB [ \-\-trace ]
.RB [ \-\-logstderr ]
.RB [ \-n ]
.RB [ \-p
.IR preffile ]

.SH DESCRIPTION
.B openipmigui
is a GUI interface using the OpenIPMI library.  It provides a
tree-structured view of the IPMI domains it connected to.

By default
.B openipmigui
starts up with no connections or anything of that nature.  You must
open connection to domains yourself.  However, you may save your
current configuration and the current domains and some GUI settings;
these will be automatically restored at startup.

The main window has a tree on the left and a log window on the right.
The log window captures informational and error logs from OpenIPMI.
Note that events are not reported in the log window, you must open an
SEL window for a domain to view the events.

.SH OPTIONS
.TP
.B \-\-dmsg
Turn on message debugging, this will dump all messages to debug log output.
.TP
.B \-\-drawmsg
Turn on raw message debugging, this will dump all low-level messages to
debug log output.  This differes from normal message debugging in that all
protocol messages are also dumped, not just IPMI messages.
.TP
.B \-\-dmem
Turn on memory debugging, this will cause memory allocation and
deallocations to be checked.  When the program terminates, it will
dump all memory that was not properly freed (leaked), and overruns,
writing of freed memory, or other types of memory errors.
.TP
.B \-\-logstderr
Send all logs to standard error output as well as the GUI console.
.TP
.B \-\-trace
Turn on python tracing so that all python procedure calls entered and
exited will be printed out.
.TP
.B \-n
Turn off reading of the preferences file at startup.
.TP
.BI \-p\  preffile
Use \fIpreffile\fP as the name of the preferences file instead of
\fB$HOME/.openipmigui.startup\fP.

.SH MENU BAR

The main menu bar contains two pull-down menus: File and View.

The File menu contains Exit, Open Domain, and Save Prefs commands.
The Exit command does as described, and it does not prompt to be sure
you want to exit.  The Open Domain command is how you establish a
connection to an IPMI system.  The Save Prefs command will store the
current domains and the configuration of the main GUI window.  These
settings will be restored at startup.  The settings are stored in
$HOME/.openipmigui.startup; this file may be removed to clear all the
settings.

The View menu contains commands to expand and contract the entire
tree.

.SH TREE VIEW

The tree window has a top-level list of all the domains for which
OpenIPMI has connections (or pending connection).  Each domain
exapands into domain-specific information and a list of entities and
Management Controllers (MCs) for that domain.

Right click drives most of the operations in the tree window.  Many
tree elements have pull-down menus driven by a right mouse button
click.  When in doubt, click.

Color is used in the GUI to denote error status, and shading is used
to denote availability.  If a sensor has an error, the tree entry for
that sensor will change colors.  Black means no error, yellow means
warning, red means critical, and blue means non-recoverable.  These
errors propigate up, so the entity containing that sensor will be the
same color as the most critical error for the sensors underneath it.
The same goes for domain, it will be the same color as the most
critical error for the entities underneath it.  This way, if you watch
the domain, you can determine if a system is in trouble with just the
top-level view, and you can drill down to find the problem.

If a tree entry is shaded, it means that the entry has information
identifying it but does not exist.  For instance, a domain with no
connection, an entity that is not present, or an MC that is not
present.

Note that all information visible in the tree is polled.  Only visible
information is polled, so if you expose something via expanding a tree
element or scrolling, it may be a few seconds before the data is
updated.

.SH DOMAINS

The domains contain various high-level information and an aggregation
of all the System Event Logs (SELs) contained within it.  The SEL
Rescan Time and IPMB Rescan Time are settable via right-click; these
values are remembered when saved and restored.  (Note that these
values in the MC-specific entries are not saved and restore, only the
domain level ones).  Right clicking on the domain gives a set of
operations, including the ability to view the SEL.

.SH ENTITIES

The Entities element under the domain element contains a list of the
entities discovered for that domain.  This is only the top-level
entities (ones not contained in other entities); entities contained in
other entities are listed under their parent entity.  This keeps the
clutter down for very complicated systems, like ATCA.

If the entity is hot-swappable, hot-swap activation and deactivation
commands are available via right-click on the entity. If the entity
has FRU information, you can view that via a right-click command, too.

The information under an entity also contains a sensors and controls
list for the entity.

.SH SENSORS

Sensors are available under the entity they belong to.  Sensor
information is beyond the scope of this document, as IPMI sensors are
very complex.  See the IPMI document from OpenIPMI for information
about sensors.

As expected, right-clicking on a sensor gives operations available on
that sensor.

.SH CONTROLS

Controls are available under the entity they belong to.  control
information is beyond the scope of this document, as IPMI controls are
very complex.  See the IPMI document from OpenIPMI for information
about controls.

As expected, right-clicking on a control gives operations available on
that control.


.SH MANAGEMENT CONTROLLERS (MCs)

An MC is the intelligent node in an IPMI system.  It is not generally
required to know about MCs for normal IPMI sensor and control
operations.  You do need access to MCs, though, for system
configuration.

An MC contains a lot of information about itself and its capabilities;
this is generally automatically interpreted by OpenIPMI, but the
information is available for view.  For instance, if an MC contains an
SEL device, OpenIPMI will automatically handle getting events from it,
but you can also view this setting under the MC information.

A number of operations are available under the right-click menu on the
MC; most of these are self-explanatory.  The MC is generally
configured via the Channel Info command under this.

.SH CONNECTIONS

A set of connections for the domain is listed under the domain.  Each
connection has one or more ports.  The state of these ports may be
viewed by color; if a port is yellow (warning), it is down.  If a
connection is red (severe), then all its ports are down.  If all
connections are down, the "Connections" tree item turns blue
(critical).

.SH CHANNEL INFO

If you right click on an MC, it will show a Channel Info entry.  This
is used to configure the LAN and channel information for the MC.  If
you select this, it will pull up a list of channels in a tree view.
Note that unlike the main tree, this information is not dynamically
updated; it will not change if the underlying data changes.

To configure individual channels, expand them.  The Info settings are
read only and cannot be changed.  The User Access settings can be
changed by right-clicking on them.  Note that the change is not
immediate, you must use the Save button at the bottom of the window to
save the changes you have made.  Also, there are two versions of the
User Access settings.  One is volatile, and will update the current
settings for the MC, but if the MC restart the non-volatile settings
will be restored.  Setting the non-volatile settings will not cause an
immediate change, but those changes will be restored on the next MC
restart.

Some channels have user information; you can configure these by
selecting that command in the right-click menu of a channel.

LAN channels also have LAN configuration; you can configure these by
selecting that command in the right-click menu of a channel.

.SH USER INFO

IPMI systems with remote access have users that can be used to log
in.  These users are configured by the User Info command in the
channel listing.  Note that some user setting (the name and password)
are global to all channels on the MC.

Due to the IPMI spec and some unfortunate implementation bugs, some
wierd issues exist with this information.  The user has an Enabled
value that tells whether the user is enabled or not.  When initially
displayed, this field shows as a "?" because this field is not
readable. It will displayed as the actual value when it is modified,
but will not be changed otherwise.

There is also a session limit field that gives the maximum number of
sessions a user is allowed to make.  This field is also not readable
and displays as a "?" when first displayed.  This field is marked as
optional in the IPMI spec, which means the field does not have to be
supplied.  Some implementations, however, require this field to be
set.  If you get a DataLenInv error back from your system when setting
something, you must set the field to something for any value to be
set.  Generally, setting it to zero disables the field and is the
right thing to do.

.SH LANPARMS

In the channel display on a 8023_LAN channel, a LANPARM configuration
command is available.  This pulls up all the parameters for the LAN
and allows them to be set.  Right-clicking on an item allows it to be
changed.

.SH PEFPARMS

If an MC support PEF, it will have a "PEF Parms" menu item in the
pulldown menu for the MC.  This pulls up all the parameters for the PEF
and allows them to be set.  Right-clicking on an item allows it to be
changed.

.SH SOLPARMS

In the channel display on a 8023_LAN channel, a LANPARM configuration
command is available if the MC supporte SOL (Serial Over LAN).  This
pulls up all the parameters for the LAN and allows them to be set.
Right-clicking on an item allows it to be changed.

.SH SOL

On a connection, an "Open SOL" menu item is present.  If the BMC on
the far end of the connection has SOL (Serial Over LAN) support, then
you can establish a serial connection with the BMC, assuming it is
configured properly and you have the proper rights.

When you choose this, you will get a new connection with a VT100/ANSI
terminal emulator.  The connection will not yet be opened, you must
open it (under the "File") to establish the connection.  When the
status at the bottom right says "connected", you are ready to type.
If something goes wrong, the connection may drop and you will have to
reopen it.  You can also close the connection from the file menu, or
quit the emulator.

The "Controls" menu for the SOL connection contains a large number of
settings for SOL.  These allow you to manipulate various value on the
remote side of the connection.  These are:

.TP
.B Accept Input
If disabled, this will cause data from the BMC to be refused,
effectively flow-controlling the data (assuming that the BMC actually
supports this properly, some don't)

.TP
.B Use Encryption
Encrypt the data on the LAN using RMCP+ encryption negotiated for the
connection.  Only available when the connection is closed, this cannot
be changed dynamically.

.TP
.B Use Authentication
Authenticate the data on the LAN using RMCP+ authentication negotiated
for the connection.  Only available when the connection is closed,
this cannot be changed dynamically.

.TP
.B Deassert CTS/DCD/DSR on connect
Don't raise the lines when a connection is made; this allows things
like baud rate to be set before the device on the far end of the
serial port knows there is a connection.

.TP
.B CTS Assertable
If set, the BMC has control over the CTS line.  Otherwise, the CTS
line is deasserted.

.TP
.B DCD/DSR Asserted
Set the DCD and DSR lines.

.TP
.B RI Asserted
Assert the Ring Indicator line, can be used for remote wakeup on some
systems (though, with IPMI, you have to wonder why you wouldn't use
IPMI for this).

.TP
.B Set Ack Timeout
Sets the timeout waiting for an ACK from the BMC.

.TP
.B Set Ack Retries
Sets the number of times a packet is retried before the connection is
closed.

.TP
.B Send Break
Send a 300ms break signal to the remote serial port.

.TP
.B Serial Rate
Sets the bitrate on the serial port.  Not all systems support this.
Some, in fact, completely ignore this setting and use 19.2K.  Note
that this is only the volatile rate.  The non-volatile rate setting is
done with SOLPARM.

.TP
.B Serial Alert Behavior
This sets what to do when an IPMI alert occurs and is going out the
serial port somehow.  If set to "succeed", the SoL session will be
disconnected when an alert comes out.

.TP
.B Queue Flush
Allows various queues to be flushed.

.SH "SEE ALSO"
.BR ipmish (1),
IPMI document from OpenIPMI

.SH "KNOWN PROBLEMS"
Some of missing functionality.  See the list in the
swig/python/openipmigui/TODO if you want to work on something.

.SH AUTHOR
.PP
Corey Minyard <cminyard@mvista.org>


ChangeLog
2006-10-16 Corey Minyard <cminyard@mvista.com>

	* doc/IPMI.ltx: Added some docs on redundant connections and on
	ATCA.

	* man/openipmicmd.1: Added information about the proper handling
	of response messages.

	* swig/python/openipmigui/gui.py: Fix a problem where sensor
	settings would get greyed

2006-10-12 Philipp Hahn

	* man/ipmi_cmdlang.7: Various cleanups
	* man/openipmi_conparms.7: ditto
	* man/openipmigui.1: ditto
	* man/openipmish.1 ditto
	* man/rmcp_ping.1 ditto

2006-10-13 Corey Minyard <cminyard@mvista.com>

	* swig/python/openipmigui/_mc_lanparm.py: Fix a problem in a
	pulldown.

2006-10-11 Corey Minyard <cminyard@mvista.com>

	* swig/python/openipmigui/_misc.py: Added an int array to hex
	string converter.

	* swig/python/gui_cmdwin.py: Print command output in hex.

	* swig/python/_sel.py: Print event data in hex.

2006-10-06 Corey Minyard <cminyard@mvista.com>

	* lib/ipmi_smi.c: Fix the IPMB address setting code to only set
	the channels that are reported to it.

	* lib/oem_atca_conn.c: Properly initialize the ipmb array that is
	passed to the other code.

2006-10-05 Corey Minyard <cminyard@mvista.com>

	* sample/ipmi_serial_bmc_emu.c, sample/Makefile.am: Added a serial
	BMC emulator.

2006-09-22 Corey Minyard <cminyard@mvista.com>

	* swig/python/openipmigui.py: Fixed a problem dealing with
	different python/TCL thread models in the GUI.  Major ugliness.
	See the code for details.

	* configure.in: Move to version 2.0.8
	
2006-09-21 Coly Li <colyli@gmail.com>

	* swig/python/Makefile.am: Fix the path mistake for install 
	handling.

2006-09-21 Coly Li <colyli@gmail.com>
	
	* swig/OpenIPMI.i: add the missed IPMI_SWIG_C_CB_EXIT to
	add_domain_change_handler and remove_domain_change_handler.

2006-09-20 Corey Minyard <cminyard@mvista.com>
	
	* cmdlang/Makefile.am: Add proper uninstall handling
	* include/OpenIPMI/Makefile.am: ditto
	* include/OpenIPMI/internal/Makefile.am: ditto
	* sample/Makefile.am: ditto
	* swig/perl/Makefile.am: ditto
	* swig/python/Makefile.am: ditto
	* swig/python/openipmigui/Makefile.am: ditto

	* swig/OpenIPMI.i: Fix a misnamed #ifdef that caused
	cleanup_lang() to not be called.

2006-09-19 Corey Minyard <cminyard@mvista.com>

	* swig/python/openipmigui/_mc_solparm.py: Use the proper data
	object for a SOL parameter window.

2006-09-15 <nok at users.sourceforge.net>

	* man/openipmicmd.1: Fix the man page number in the header.

2006-09-13 Corey Minyard <cminyard@mvista.com>

	* lib/mc.c: Don't set the event receiver in the MC recheck if the
	user has asked it to not be set.

2006-08-24 Corey Minyard <cminyard@mvista.com>

	* swig/python/openipmigui/gui_list.py: Split up the list processing
	into two parts, one for a standalone toplevel list, and one for a
	list that can be contained in something else.

2006-08-01 Corey Minyard <cminyard@mvista.com>

	* configure.in, doc/Makefile.am: Autodetect the dia version and use
	the right command-line arguments depending on version.

	* Makefile.am, swig/Makefile.am: Add DIST_SUBDIRS so "make dist"
	will work even if you don't have all the tools present.

	* Makefile.am, include/OpenIPMI/Makefile.am: Added proper files to
	DISTCLEANFILES.

2006-07-31 Corey Minyard <cminyard@mvista.com>

	* lib/entity.c: Fixed a problem with error handling for bad SDRs.

2006-07-24 Corey Minyard <cminyard@mvista.com>

	* FAQ - Added info on interpreting errors.

2006-07-22 Corey Minyard <cminyard@mvista.com>

	* lib/rakp.c: Fix a problem that causes a crash in RMCP+ if the
	IPMI interface is deleted between the RAKP1 and RAKP3 messages.

	* swig/python/openipmigui/_mc_chan.py: Fix a shutdown problem that
	could cause wierd things to happen if you delete the channel
	window while it was being fetched.

2006-07-21 Gerhard Obrecht <varnhalt2 at users.sourceforge.net>

	* lib/oem_kontron_conn.c: Add more board support and fix
	some IPMB addressing problems.

2006-07-18 Corey Minyard <cminyard@mvista.com>

	* swig/python/openipmigui.py, swig/python/openipmigui/gui.py: Fix
	some problems with shutdown, gets rid of an annoying failure at
	shutdown and fixes a lockup problem if there are no domains.

2006-07-16 Corey Minyard <cminyard@mvista.com>

	* sample/sample.c: Change over to use ipmi_parse_args2() so we
	are setting a good example for the users.
	* sample/sample2.c: ditto
	* sample/sample3.c: ditto

2006-06-27 Corey Minyard <cminyard@mvista.com>

	* swig/python/openipmigui/gui.py: Fix cleanup of items to clear
	out parents properly.

2006-06-23 Gerhard Obrecht <varnhalt2@users.sourceforge.net>

	* lib/oem_kontron_conn.c: Add new board support.

2006-06-18 Corey Minyard <cminyard@mvista.com>

	* FAQ, README: Add info about the encryption used by OpenIPMI.

2006-06-08 Corey Minyard <cminyard@mvista.com>

	* configure.in: Convert wxpython into Tkinter.

	* configure.in: Older version of Tcl don't have Tcl_GetTime, they
	have an equivalent TclpGetTime (that was private, but everyone
	still used it).

	* README - Update some information, including adding all the
	required and optional packages.

	* OpenIPMI.spec.in: Added the Tcl stuff.

	* Makefile.am: Added CONFIGURING_FOR_LAN

	* configure.in: Move to version 2.0.7.

2006-06-07 Corey Minyard <cminyard@mvista.com>

	* swig/python/openipmigui/*: Get SoL and the command window working
	with Tix, the GUI conversion is complete except for testing.

	* doc/Makefile.am: Convert the export option for dia from
	--export-to-format to --filter.

2006-06-02 Corey Minyard <cminyard@mvista.com>

	* lib/pefc.c: Only call the set_done() callback if it is non-NULL.

	* lib/mc.c: If a set_user() call is done with nothing changed in
	the user, make sure the handler is not NULL before calling and
	call the right free function.

	* swig/python/openipmigui*: Rewrote the GUI using Tk.  The SOL
	terminal is not done yet, but I think everything else works.
	Still needs testing, but most stuff should work ok.

2006-06-01 Corey Minyard <cminyard@mvista.com>


swig/python/openipmigui.py
#!/usr/bin/env python

# openipmigui.py
#
# The openipmi GUI startup file
#
# Author: MontaVista Software, Inc.
#         Corey Minyard <minyard@mvista.com>
#         source@mvista.com
#
# Copyright 2005 MontaVista Software Inc.
#
#  This program is free software; you can redistribute it and/or
#  modify it under the terms of the GNU Lesser General Public License
#  as published by the Free Software Foundation; either version 2 of
#  the License, or (at your option) any later version.
#
#
#  THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED
#  WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
#  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
#  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
#  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
#  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
#  OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
#  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
#  TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
#  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
#  You should have received a copy of the GNU Lesser General Public
#  License along with this program; if not, write to the Free
#  Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#
#
# This is a GUI interface to the OpenIPMI library.
#
# Naming Conventions:
#
#  The following names are used universally throughout this program.
#  Note that you have two versions of each object, the Python object
#  (class is defined in this program) and the OpenIPMI version of the
#  object (defined by the OpenIPMI library).
#
#  domain - OpenIPMI.ipmi_domain_t
#  domain_id - OpenIPMI.ipmi_domain_id_t
#  entity - OpenIPMI.ipmi_entity_t
#  entity_id - OpenIPMI.ipmi_entity_id_t
#  mc - OpenIPMI.ipmi_mc_t
#  mc_id - OpenIPMI.ipmi_mc_id_t
#  sensor - OpenIPMI.ipmi_sensor_t
#  sensor_id - OpenIPMI.ipmi_sensor_id_t
#  control - OpenIPMI.ipmi_control_t
#  control_id - OpenIPMI.ipmi_control_id_t
#
#  d - Domain
#  e - Entity
#  m - MC
#  s - Sensor
#  c - Control

import os
import Tix
import sys
import OpenIPMI
from openipmigui import _domain
from openipmigui import gui
from openipmigui import _saveprefs
from openipmigui import gui_cmdwin

# Used to enable internal debug output
verbosity = 0

#
# Nasty thread support stuff.
#
# If python is threaded but TCL is not, we have ugliness to deal with.
# In this case, the _tkinter code has a tcl_lock that is claimed
# whenever it enters TCL code, including the event loop.  If we run
# OpenIPMI events inside the TCL event loop, it will call directly
# into the OpenIPMI calls and bypass _tkinter, leaving the lock held.
# Inside OpenIPMI, it might call back to the python code which can
# then do a TCL call, which will try to claim tcl_lock and deadlock.
#
# We fix this in this case by having a separate thread
# (openipmi_driver() below) run OpenIPMI events, and not in the TCL
# event loop.
#
# If both python and TCL are threaded, the tcl_lock is not used and
# everything is fine with a single thread.  If python is not threaded,
# then threads don't matter.  In those cases we use the TCL event loop
# for everything.
#
# See the _tkinter.c file in the python distribution for more details.
#
shutdown_thread = False
def openipmi_driver():
    while (not shutdown_thread):
        OpenIPMI.wait_io(1000)
        pass
    return

class TopHandler(Tix.Tk):
    def __init__(self, preffile, log_file, histfile):

        Tix.Tk.__init__(self)
        
        self.domains = { };
        self.preffile = preffile
        self.log_file = log_file
        self.histfile = histfile
        return

    def domain_change_cb(self, op, domain):
        if (op == "added"):
            self.domains[domain.get_name()].connected(domain)
        elif (op == "removed"):
            self.domains[domain.get_name()].remove()
            pass
        return

    def SetUI(self, ui):
        self.ui = ui;
        return

    def savePrefs(self):
        objs = self.domains.values()
        objs.append(self.ui)
        _saveprefs.save(objs, self.preffile)
        return

    def log(self, level, log):
        if (self.log_file != None):
            self.log_file.write(level + ": " + log + "\n")
        self.ui.new_log(level + ": " + log)
        return

    def quit(self):
        global shutdown_thread
        shutdown_thread = True;
        gui_cmdwin._HistorySave(self.histfile)

        OpenIPMI.set_log_handler(DummyLogHandler())
        OpenIPMI.shutdown_everything()
        if (self.debug_mem):
            print "OpenIPMI is shutdown, memory problems (SEGVs) after this"
            print " are likely due to OpenIPMI data not being freed until"
            print " after this point due to the python garbage collector"
            pass
        sys.exit()
        return

class DummyLogHandler:
    def __init__(self):
        pass

    def log(self, level, log):
        sys.stderr.write(level + ": " + log + "\n")

class CmdlangEventHandler:
    def __init__(self, app):
        self.app = app
        return
    
    def cmdlang_event(self, event):
        if (not self.app.ui.logevents):
            return
        name = [ "" ]
        value = [ "" ]
        vtype = [ "" ]
        level = [ 0 ]
        event.restart()
        estr = "Event:"
        more = event.next_field(level, vtype, name, value)
        while (more != 0):
            estr += "\n  %*s%s: %s" % (level[0], "", name[0], value[0])
            more = event.next_field(level, vtype, name, value)
            pass
        self.app.ui.new_log(estr)
        return
    
    pass

def trace(frame, event, arg):
    print (event + ": " + frame.f_code.co_name +
           "(" + frame.f_code.co_filename + ":" + str(frame.f_lineno) + ")")
    return trace

def run(args):
    global verbosity

    preffile = os.path.join(os.environ['HOME'], '.openipmigui.startup')
    histfile = os.path.join(os.environ['HOME'], '.openipmigui.history')

    debug_msg = False
    debug_rawmsg = False
    debug_mem = False
    do_trace = False
    read_preffile = True
    log_file = None

    # Skip program name.
    carg = 1

    while (carg < len(args)):
        arg = args[carg]
        carg += 1
        if (arg == "--dmsg"):
            debug_msg = True
        elif (arg == "--drawmsg"):
            debug_msg = True
        elif (arg == "--dmem"):
            debug_mem = True
        elif (arg == "--verbose"):
            verbosity += 1
        elif (arg == "--trace"):
            do_trace = True
        elif (arg == "--logstderr"):
            log_file = sys.stderr
        elif (arg == "--logstdout"):
            log_file = sys.stdout
        elif (arg == "-n"):
            read_preffile = False
        elif (arg == '-p'):
            if (len(args) == 0):
                print "No argument given for -p";
                return
            preffile = args[carg]
            carg += 1
        else:
            print "Unknown argument: " + arg
            return
        pass

    if (debug_mem):
        OpenIPMI.enable_debug_malloc()
        pass

    top = TopHandler(preffile, log_file, histfile)

    # Detect if we need a separate OpenIPMI driver thread.  See the
    # openipmi_driver function above for the reason.
    try:
        import thread
        try:
            top.tk.getvar("tcl_platform", "threaded")
            # Tcl is threaded, no need for another thread.
            need_separate_openipmi_thread = False
        except:
            # Python is threaded, but Tcl is not.  Need to run the
            # OpenIPMI event loop in another thread.
            need_separate_openipmi_thread = True
            pass
        pass
    except:
        # No thread support, can't use another thread.
        need_separate_openipmi_thread = False
        pass

    if (need_separate_openipmi_thread):
        if (verbosity >= 1):
            print "Creating separate OpenIPMI event driver thread"
            pass
        OpenIPMI.init()
        thread.start_new_thread(openipmi_driver, ())
        pass
    else:
        if (verbosity >= 1):
            print "Using TCL event loop, no threads"
            pass
        OpenIPMI.init_tcl()
        pass

    if (debug_rawmsg):
        OpenIPMI.enable_debug_rawmsg()
        pass
    if (debug_msg):
        OpenIPMI.enable_debug_msg()
        pass
    if (do_trace):
        sys.settrace(trace)
        pass

    if (read_preffile):
        _saveprefs.restore(preffile)
    gui_cmdwin._HistoryRestore(histfile)
    
    mainhandler = top

    OpenIPMI.add_domain_change_handler(_domain.DomainWatcher(mainhandler))

    mainhandler.debug_mem = debug_mem
    top.title('OpenIPMI GUI')

    ui = gui.IPMIGUI(top, mainhandler)
    mainhandler.SetUI(ui)

    OpenIPMI.add_domain_change_handler(mainhandler)
    OpenIPMI.set_log_handler(mainhandler)

    _domain.RestoreDomains(mainhandler)

    OpenIPMI.set_cmdlang_event_handler(CmdlangEventHandler(mainhandler))

    top.mainloop()

    mainhandler.quit()


if __name__ == "__main__":
    run(sys.argv)
